app/models/metric/ci_mixin/capture.rb

```diff
@@ -175,6 +175,7 @@ def perf_capture(interval_name, start_time = nil, end_time = nil)
 
     start_range = end_range = counters = counter_values = nil
     _, t = Benchmark.realtime_block(:total_time) do
+      # TODO why we call capture here? We call the same in processing.
```

Comment:
You mean via `vim_performance_state_for_ts`?  It looks like that's only in the event that we are missing the state:
```
state = vim_performance_states.find_by(:timestamp => ts)
...
state ||= perf_capture_state
``

So this the the primary place it should be called from.  Maybe we can get rid of the "backup" in perf_process?

Reply:
yeah possibly, my tests are a bit weird with this, since I do run capture 6 days back, most of the vim_performance_states are missing. So then it does a lot of duplicate queries trying to catch them, ending with assigning the last vim_performance_states to every ts.

Reply:
I'm not thrilled with the `ts` lookup caching. There may be a way to explicitly look them up and populate them. I'm suspect this code is causing a memory leak that @NickLaMuro is researching.  But also, it would be nice to load the correct `ts` values for @Ladas .

(Sorry @NickLaMuro - I may have misunderstood and unnecessarily roped you in)

Reply:
Hum, is it? I was thinking it could mem leak, but it just stores cached values in instance variable? (or is it doing something nasty inside?)

@kbrock right, so I am at least calling preload_vim_performance_state_for_ts_iso8601, so it doesn't do n+1 queries, but it still does n+1 queries if it can't find some hour state. But yeah, more  refactoring can be done here, but perf-wise, this is not bottleneck now.